### PR TITLE
Set PHP_INI_SCAN_DIR only if not set

### PR DESCRIPTION
**Pull request by lstrojny**
Uses the preferred binary wrapper and only sets PHP_INI_SCAN_DIR if it isn’t set yet. This is so that users can provide there own PHP_INI_SCAN_DIR, e.g. to restart the current PHP process without xdebug.
E.g. composer/xdebug-handler uses this mechanism to turn off xdebug (https://github.com/composer/xdebug-handler/blob/cf19e3380ffd84cea4a2cbfcc599ea07ef01f18a/src/XdebugHandler.php#L441)

###### Things done

- Built on platform(s)
  - [x] aarch64-darwin
  - [x] Verified that https://github.com/composer/xdebug-handler works afterwards

#### Reproduction

Given this script with `composer/xdebug-handler` installed:

```php
<?php
include 'vendor/autoload.php';

var_dump('XDEBUG LOADED', extension_loaded('xdebug'));

$handler = new \Composer\XdebugHandler\XdebugHandler('repro');
$handler->setPersistent();
$handler->check();
```

#### Before the change
```
string(13) "XDEBUG LOADED"
bool(true)
Cannot load Zend OPcache - it was already loaded

Warning: Module "mbstring" is already loaded in Unknown on line 0

Warning: Module "filter" is already loaded in Unknown on line 0

Warning: Module "intl" is already loaded in Unknown on line 0

Warning: Module "openssl" is already loaded in Unknown on line 0

Warning: Module "dom" is already loaded in Unknown on line 0

Warning: Module "tokenizer" is already loaded in Unknown on line 0

Warning: Module "xmlwriter" is already loaded in Unknown on line 0

Warning: Module "SimpleXML" is already loaded in Unknown on line 0

Warning: Module "curl" is already loaded in Unknown on line 0

Warning: Module "zlib" is already loaded in Unknown on line 0

Warning: Module "ctype" is already loaded in Unknown on line 0
string(13) "XDEBUG LOADED"
bool(true)
```

#### After the change
```
string(13) "XDEBUG LOADED"
bool(true)
string(13) "XDEBUG LOADED"
bool(false)
```

### CODE REVIEW

File: pkgs/development/interpreters/php/generic.nix
Comment (jtojnar) [inline pkgs/development/interpreters/php/generic.nix:173]: Previous attempt https://github.com/NixOS/nixpkgs/pull/157287.

IIRC, the issue with `--set-default` is that it would break other extensions.

File: pkgs/development/interpreters/php/generic.nix
Comment (jtojnar) [inline pkgs/development/interpreters/php/generic.nix:144]: Is there behavioural difference with a binary wrapper? If not perhaps split it into a separate commit with its own rationale.

**Review by drupol — APPROVED** [review]
LGTM

File: pkgs/development/interpreters/php/generic.nix
Comment (jtojnar) [inline pkgs/development/interpreters/php/generic.nix:1]: Would be nice to have integration tests with projects affected by this in Nixpkgs to make sure it does not break anything. (Probably will need to commit a `composer2nix`-generated code into Nixpkgs.)

**Comment by drupol** [comment]
Glad to see you around Lars !

File: pkgs/development/interpreters/php/generic.nix
Comment (lstrojny) [reply pkgs/development/interpreters/php/generic.nix:173]: Yup, and that is the intended behavior. If I specify a custom php.UBI, I start with a blank slate

File: pkgs/development/interpreters/php/generic.nix
Comment (lstrojny) [reply pkgs/development/interpreters/php/generic.nix:144]: No, there shouldn't be a difference for the particular case. Using the binary wrapper allows using it in a shebang on darwin because execvp doesn't like shell scripts. Do you want me to split the commits?

File: pkgs/development/interpreters/php/generic.nix
Comment (lstrojny) [reply pkgs/development/interpreters/php/generic.nix:1]: What projects are you thinking about? 

Pretty sure adding tests is out of my depth though

File: pkgs/development/interpreters/php/generic.nix
Comment (jtojnar) [reply pkgs/development/interpreters/php/generic.nix:144]: Yeah, since it is unrelated to this change, a separate commit with explanation would be better.

File: pkgs/development/interpreters/php/generic.nix
Comment (jtojnar) [reply pkgs/development/interpreters/php/generic.nix:173]: Well, the Nixpkgs’s PHP infrastructure relies on this to enable even the default extensions like PDO or session so without this, it will not have almost any.

Specifically, when it [sets `PHP_INI_SCAN_DIR` to empty string](https://github.com/composer/xdebug-handler/blob/cf19e3380ffd84cea4a2cbfcc599ea07ef01f18a/src/XdebugHandler.php#L443), wrapper would [ignore the default value provided with `--set-default` ](https://github.com/NixOS/nixpkgs/blob/5911be300ddffbb55c9f7af2a06132c2de61acf1/pkgs/build-support/setup-hooks/make-wrapper.sh#L130).

But maybe we should not use `PHP_INI_SCAN_DIR` to control extensions and instead patch PHP to also respect `NIX_PHP_INI_SCAN_DIR` or `NIX_PHP_ENABLE_EXTENSION` which the packages like xdebug-handler would leave alone.

Comment (lstrojny) [reply pkgs/development/interpreters/php/generic.nix:173]: Yup, that is exactly intended. If I provide a custom ini path it's my duty to load the extensions I want. Case in point, `composer/xdebug-handler` will replicate all but the xdebug config when using the persistent option

File: pkgs/development/interpreters/php/generic.nix
Comment (jtojnar) [reply pkgs/development/interpreters/php/generic.nix:1]: Looks like PHPStan mentioned in the [previous PR](https://github.com/NixOS/nixpkgs/pull/157287) also [uses XdebugHandler](https://github.com/phpstan/phpstan-src/blob/44e0d9a87798c940a9abeaa2b6fe6d0b04611f16/src/Command/CommandHelper.php#L103-L106) so based on https://gist.github.com/guillemcanal/92f4a97bb31d25522a9a7658a5ff5019, maybe something like:

```
<?php

require __DIR__ . '/vendor/autoload.php';

// Most PHP CLI tools run this snippet to disable Xdebug at runtime
use Composer\XdebugHandler\XdebugHandler;

assert(extension_loaded('xdebug'), 'Xdebug extension should be available before running XdebugHandler.');

$xdebug = new XdebugHandler('myCliApp');
$xdebug->setPersistent();
$xdebug->check();
unset($xdebug);

assert(!function_exists('xdebug_info') || in_array('off', xdebug_info( 'mode' ), true), 'Xdebug should be disabled after running XdebugHandler.');
assert(extension_loaded('pdo'), 'PDO extension should be still available after running XdebugHandler.');
```

File: pkgs/development/interpreters/php/generic.nix
Comment (jtojnar) [reply pkgs/development/interpreters/php/generic.nix:173]: But I am **not** providing custom ini path, Nixpkgs is doing it (transparently to the user).

Comment (lstrojny) [reply pkgs/development/interpreters/php/generic.nix:173]: Just to make sure we are not talking past each other, let me restate my understanding of the situation.

The `php` executable in nix is a wrapper script that sets `PHP_INI_SCAN_DIR` so that nix can load a custom php.ini at a non-default location. That stock php.ini loads whatever extensions are available.

Because there are tools out there, that set `PHP_INI_SCAN_DIR` to a custom location. What those tools are attempting to achieve is to be able to load a completely different PHP configuration and make that persistent for all sub processes.

Now, the whole point is, that once a custom ini location is specified, all stock configuration from nix is gone and instead (not additionally!) the custom configuration is loaded. That’s exactly the same behavior as if nix would not set the stock nix configuration via `PHP_INI_SCAN_DIR` but instead use the built-in php.ini location.

```
php --info | grep -E "Additional .ini files|Configuration File \(php.ini\)"
Configuration File (php.ini) Path => /nix/store/s7rgnbhfbrbli6py2f726hd89ch54r73-php-8.2.3/lib
Additional .ini files parsed => /nix/store/vn7h4va52s0gbhys2chiwdkr3rxnzvz1-php-with-extensions-8.2.3/lib/php.ini
```

```
/nix/store/vn7h4va52s0gbhys2chiwdkr3rxnzvz1-php-with-extensions-8.2.3/bin/.php-wrapped --info | grep -E "Additional .ini files|Configuration File \(php.ini\)"
Configuration File (php.ini) Path => /nix/store/s7rgnbhfbrbli6py2f726hd89ch54r73-php-8.2.3/lib
Additional .ini files parsed => (none)
```

That’s why using `--set-default` is the right choice. It behaves symmetrically to `PHP_INI_SCAN_DIR` vs. built-in location and is exactly what any tool would expect that tinkers with `PHP_INI_SCAN_DIR`.

What `composer/xdebug-handler` does under the hood is to copy all of the current configuration except the settings that enable xdebug. Conceptually, one could implement functionality that does something similar but for opcache so that the current process is restarting without opcache. Or without PDO, or xmlreader or whatever. 

Point is: dropping all stock configuration is the right thing, as it’s the job of the tool building upon `PHP_INI_SCAN_DIR` to reconfigure PHP in whatever way desired.

File: pkgs/development/interpreters/php/generic.nix
Comment (lstrojny) [reply pkgs/development/interpreters/php/generic.nix]: Reverted that change so that this PR can get it, will open a 2nd issue once this has landed.

File: pkgs/development/interpreters/php/generic.nix
Comment (jtojnar) [reply pkgs/development/interpreters/php/generic.nix:173]: > That stock php.ini loads whatever extensions are available.

The issue there is no stock `php.ini`:

```ShellSession
$ php --info | grep -E 'Configuration File|\.ini files'
Configuration File (php.ini) Path => /nix/store/9dr0yiky45hnrvgkk2gybijxbssm7b80-php-8.2.3/lib
Loaded Configuration File => (none)
Scan this dir for additional .ini files => /nix/store/b7nhlzic3y18ckngxg3zxjqnicjbqajd-php-with-extensions-8.2.3/lib
Additional .ini files parsed => /nix/store/b7nhlzic3y18ckngxg3zxjqnicjbqajd-php-with-extensions-8.2.3/lib/php.ini
```

Compare with Fedora:

```ShellSession
$ php --info | grep -E 'Configuration File|\.ini files'
Configuration File (php.ini) Path => /etc
Loaded Configuration File => /etc/php.ini
Scan this dir for additional .ini files => /etc/php.d
Additional .ini files parsed => /etc/php.d/10-opcache.ini,
$ PHP_INI_SCAN_DIR= php --info | grep -E 'Configuration File|\.ini files'
Configuration File (php.ini) Path => /etc
Loaded Configuration File => /etc/php.ini
Scan this dir for additional .ini files => (none)
Additional .ini files parsed => (none)
```

And since because

1. Nix does not allow dependency cycles between packages
2. There is no central location where PHP could look for extensions
3. Nixpkgs builds extensions as separate packages

Nixpkgs needs some method of passing the extensions to php. We have decided to use `PHP_INI_SCAN_DIR`.

> Now, the whole point is, that once a custom ini location is specified, all stock configuration from nix is gone and instead (not additionally!) the custom configuration is loaded. That’s exactly the same behavior as if nix would not set the stock nix configuration via `PHP_INI_SCAN_DIR` but instead use the built-in php.ini location.

Second issue, Nixpkgs makes this worse since it goes further in decoupling modules from core PHP package:

```ShellSession
$ PHP_INI_SCAN_DIR= php -m
[PHP Modules]
Core
date
hash
json
libxml
pcre
Phar
random
Reflection
SPL
standard
xml

[Zend Modules]
```

Again, contrast this with Fedora:

```ShellSession
$ PHP_INI_SCAN_DIR= php -m
[PHP Modules]
Core
date
filter
hash
json
libxml
openssl
pcntl
pcre
readline
Reflection
session
SPL
standard
zlib

[Zend Modules]
```

> Point is: dropping all stock configuration is the right thing, as it’s the job of the tool building upon `PHP_INI_SCAN_DIR` to reconfigure PHP in whatever way desired.

And third issue, if you drop the `ini` file from the wrapped package, you will not be able to re-add the extensions since there is no [`extension_dir`](https://www.php.net/manual/en/ini.core.php#ini.extension-dir) config.

So perhaps my point is that Nixpkgs’s wrapper is unsuitable for any use case that touches `PHP_INI_SCAN_DIR` and this PR just trades one breakage for another.

I guess an optimal solution would be patching PHP to accept `NIX_PHP_CONFIG_FILE` and then have the wrapper use that instead of relying on `PHP_INI_SCAN_DIR`. We could make `php.buildEnv` do a `symlinkJoin` of all enabled extensions and pass it through `extension_dir` config option in the wrapper.

Comment (jtojnar) [reply pkgs/development/interpreters/php/generic.nix:173]: > What composer/xdebug-handler does under the hood is to copy all of the current configuration except the settings that enable xdebug.

I missed that xdebug-handler does this when skimming the code. If every major tool uses xdebug-handler or implements similar feature, it should not be such a problem then.

Since
- the [original change](https://github.com/NixOS/nixpkgs/pull/125405) have been reverted due to issues with duplicates but that should not happen with `--set-default`,
- the [previous attempt](https://github.com/NixOS/nixpkgs/pull/157287) just appears to have stalled (maybe due to lack of confidence), and
- the other issue I vaguely recalled was actually caused by the tool running `php` with `-n` flag

I retract my objection.

**Comment by jtojnar** [comment]
@ofborg eval

**Review by jtojnar — APPROVED** [review]
I ran CI of two of my PHP projects with this branch without issue. There might still be projects or tools that this breaks but they should probably replicate what XdebugHandler does anyway. So I propose we merge this and see if anyone complains.

**Review by drupol — APPROVED** [review]

**Review by drupol — APPROVED** [review]

**Comment by etu** [comment]
So should we give this another go? I'm still scared from the last time :joy:

**Comment by lstrojny** [comment]
This is fine

**Pull request merged**
